pkg/async/cond.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Cond.go provides a context respecting sync condition
+
+package async
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// Cond is a sync.Cond that respects context cancellation.
+// It provides equivalent functionality to sync.Cond (excepting there is no `Signal` method), except that
+// the Wait method exits with error if the context cancels.
+//
+// It also provides WaitForCondition, which intends to encapsulate the common pattern of acquiring a lock,
+// checking a condition, and releasing the lock before waiting for a state change if the condition is not met.
+type Cond struct {
+	pointer atomic.Pointer[chan struct{}]
+	Mu      sync.Mutex
+}
+
+// ch returns the channel that Waiters are waiting on,
+// possibly creating one if it hasn't been initialized
+func (c *Cond) ch() chan struct{} {
+	// non atomic check for nil channel
+	load := c.pointer.Load()
+	if load == nil {
+		t := make(chan struct{})
+		c.pointer.CompareAndSwap(nil, &t)
+		return t
+	}
+	return *load
+}
+
+// Wait waits for the state change Broadcast until context ends.
+func (c *Cond) Wait(ctx context.Context) error {
+	ch := c.ch()
+	select {
+	case <-ch:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Broadcast signals the state change to all Waiters
+func (c *Cond) Broadcast() {
+	// swap in a new channel, close the old one
+	newChan := make(chan struct{})
+	ch := c.pointer.Swap(&newChan)
+	if ch != nil {
+		close(*ch)
+	}
+}
+
+// WaitForCondition acquires Cond's lock, then checks if the condition is true. If the condition is not true,
+// or the lock was not available, it releases the locker and waits for the state change Broadcast.
+// If the context ends during any of these operations, the context error is returned.
+//
+// WaitForCondition returns an unlock function that should always be called to unlock the locker.
+// unlock is safe to call regardless  of error.
+// Error should only be returned if the context ends before the condition is met.
+//
+// If it returns without error, it also locks the provided locker and the caller must call the returned function
+// to unlock it. Until they call unlock, the state should not be changed.
+func (c *Cond) WaitForCondition(ctx context.Context, condition func() bool) (unlock func(),
```

Comment:
I don't understand the addition of this function. Given the purpose of this package is to "provides a context respecting sync condition", I think we shouldn't deviate from the standard library signature in this package.

We should probably put this in a different package than `async`, maybe `pkg/syncctx`? Or something similar?

Reply:
To be clear, my issue here is matching the `Cond` signature and adding new functionality. If we want to make this a special package, I'm OK with that, but we should make it clear that our goal is not to provide the stdlib.

Reply:
I have no problem moving it. 

This function is supposed to encapsulate their instructions on how to use cond, basically you have to invoke wait in a loop, locking when you start the loop, waiting for Wait to return, and then doing your operation before unlocking. That's enough machinery to warrant encapsulation imho, and the lack of that makes it harder to use sync.Cond.

Reply:
So if you want me to move this function specifically I'm happy to. Maybe just to another file or moving the receiver to a function argument would be sufficient?

Reply:
I prefer the latter. But also saying that we're not trying to perfectly match `sync.Cond` is also fine! A comment on the `Cond` struct would suffice for me.

Reply:
I'll comment cond then

Reply:
done